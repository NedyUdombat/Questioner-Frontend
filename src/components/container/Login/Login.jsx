import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { Loader, Dimmer } from 'semantic-ui-react';

// actions
import { authenticate } from '../../../store/modules/auth';

// components
import Input from '../../presentational/Input/Input';
import Button from '../../presentational/Button/Button';

export class Login extends React.Component {
  state = {
    userData: {
      email: '',
      password: '',
    },
  };

  userInputHandler = event => {
    this.setState({
      userData: {
        ...this.state.userData,
        [event.target.name]: event.target.value,
      },
    });
  };

  submitHandler = async event => {
    event.preventDefault();
    const { userData } = this.state;
    this.props.authenticate({
      ...userData,
    });

    this.setState({
      userData: {
        ...userData,
        password: '',
      },
    });
  };

  render() {
    const { isLoading } = this.props;
    return (
      <div>
        <p
          className="f-50 text-center font-title"
          style={{
            marginBottom: '0.625rem',
          }}
        >
          Login
        </p>
        <div className="auth-card">
          <form onSubmit={this.submitHandler} autoComplete="on">
            {isLoading && (
              <Dimmer active>
                <Loader size="large">Loading</Loader>
              </Dimmer>
            )}
            <div
              className="form-group"
              style={{
                display: 'block',
              }}
            >
              <div
                className="input-div d-flex justify-content-center"
                style={{
                  marginBottom: '1rem',
                }}
              >
                <Input
                  onChange={this.userInputHandler}
                  className="form-control-input"
                  value={this.state.userData.email}
                  placeholder="Email"
                  name="email"
                  type="email"
                  required="required"
                />
              </div>
              <div className="input-div d-flex justify-content-center">
                <Input
                  onChange={this.userInputHandler}
                  className="form-control-input"
                  value={this.state.userData.password}
                  placeholder="Password"
                  name="password"
                  type="password"
                  pattern="(?=.*[a-z]).{6,}"
                  title="Must be 6 characters or more and contain at least lowercase letter"
                  required="required"
                />
              </div>
            </div>
            <div
              style={{
                marginTop: '.9235rem',
              }}
            >
              <Button type="submit" value="WELCOME BACK" className="btn-dark" />
            </div>
            <p
              className="f-20 text-grey text-center"
              style={{
                marginTop: '.9235rem',
              }}
            >
              Don&apos;t have an account? Register{' '}
              <Link to="/register" className="text-blue">
                here
              </Link>
            </p>
          </form>
        </div>
      </div>
    );
  }
}

Login.propTypes = {
  authenticate: PropTypes.func,
  isLoading: PropTypes.bool,
};

export const mapStateToProps = state => ({
  isLoading: state.auth.isLoading,
});

export default connect(
  mapStateToProps,
  { authenticate },
)(Login);
