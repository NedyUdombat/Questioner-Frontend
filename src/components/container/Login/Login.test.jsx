import React from 'react';
import { mount } from 'enzyme/build';
import { BrowserRouter } from 'react-router-dom';

import { Login, mapStateToProps } from './Login';

const props = {
  isLoading: false,
  authenticate: jest.fn(),
};

const loadingProps = {
  isLoading: true,
  authenticate: jest.fn(),
};

const mountWrapper = mount(
  <BrowserRouter>
    <Login {...props} />
  </BrowserRouter>,
);

describe('Register', () => {
  it('should render correctly', () => {
    const component = mount(
      <BrowserRouter>
        <Login {...props} />
      </BrowserRouter>,
    );
    expect(component).toMatchSnapshot();
  });

  it('should render correctly with loader', () => {
    const component = mount(
      <BrowserRouter>
        <Login {...loadingProps} />
      </BrowserRouter>,
    );
    expect(component).toMatchSnapshot();
  });

  it('mounts correctly', () => {
    const DEFAULT_STATE = {
      auth: {
        isLoading: false,
      },
    };
    const action = mapStateToProps(DEFAULT_STATE);
    expect(action).toEqual(DEFAULT_STATE.auth);
  });

  it('correctly initializes the value of the username field to an empty string', () => {
    expect(mountWrapper.find('input[name="email"]').prop('value')).toBe('');
    expect(mountWrapper.find('input[name="password"]').prop('value')).toBe('');
  });

  it('should submit the form', () => {
    const email = mountWrapper.find('input[name="email"]');
    email.instance().value = 'nedyudombat@gmail.com';
    email.simulate('change');
    const password = mountWrapper.find('input[name="password"]');
    password.instance().value = 'Iamtheadmin';
    password.simulate('change');
    const form = mountWrapper.find('form');
    form.simulate('submit', { preventDefault: jest.fn() });
    expect(mountWrapper.state()).toEqual(null);
  });
});
