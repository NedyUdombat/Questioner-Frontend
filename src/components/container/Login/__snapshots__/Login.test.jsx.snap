// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register should render correctly 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Login
      authenticate={[MockFunction]}
      isLoading={false}
    >
      <div>
        <p
          className="f-50 text-center font-title"
          style={
            Object {
              "marginBottom": "0.625rem",
            }
          }
        >
          Login
        </p>
        <div
          className="auth-card"
        >
          <form
            autoComplete="on"
            onSubmit={[Function]}
          >
            <div
              className="form-group"
              style={
                Object {
                  "display": "block",
                }
              }
            >
              <div
                className="input-div d-flex justify-content-center"
                style={
                  Object {
                    "marginBottom": "1rem",
                  }
                }
              >
                <Input
                  className="form-control-input"
                  name="email"
                  onChange={[Function]}
                  placeholder="Email"
                  required="required"
                  type="email"
                  value=""
                >
                  <input
                    className="form-control-input"
                    name="email"
                    onChange={[Function]}
                    placeholder="Email"
                    required="required"
                    type="email"
                    value=""
                  />
                </Input>
              </div>
              <div
                className="input-div d-flex justify-content-center"
              >
                <Input
                  className="form-control-input"
                  name="password"
                  onChange={[Function]}
                  pattern="(?=.*[a-z]).{6,}"
                  placeholder="Password"
                  required="required"
                  title="Must be 6 characters or more and contain at least lowercase letter"
                  type="password"
                  value=""
                >
                  <input
                    className="form-control-input"
                    name="password"
                    onChange={[Function]}
                    pattern="(?=.*[a-z]).{6,}"
                    placeholder="Password"
                    required="required"
                    title="Must be 6 characters or more and contain at least lowercase letter"
                    type="password"
                    value=""
                  />
                </Input>
              </div>
            </div>
            <div
              style={
                Object {
                  "marginTop": ".9235rem",
                }
              }
            >
              <Button
                className="btn-dark"
                type="submit"
                value="WELCOME BACK"
              >
                <button
                  className="btn btn-dark"
                  type="submit"
                >
                  WELCOME BACK
                </button>
              </Button>
            </div>
            <p
              className="f-20 text-grey text-center"
              style={
                Object {
                  "marginTop": ".9235rem",
                }
              }
            >
              Don't have an account? Register
               
              <Link
                className="text-blue"
                to="/register"
              >
                <a
                  className="text-blue"
                  href="/register"
                  onClick={[Function]}
                >
                  here
                </a>
              </Link>
            </p>
          </form>
        </div>
      </div>
    </Login>
  </Router>
</BrowserRouter>
`;

exports[`Register should render correctly with loader 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Login
      authenticate={[MockFunction]}
      isLoading={true}
    >
      <div>
        <p
          className="f-50 text-center font-title"
          style={
            Object {
              "marginBottom": "0.625rem",
            }
          }
        >
          Login
        </p>
        <div
          className="auth-card"
        >
          <form
            autoComplete="on"
            onSubmit={[Function]}
          >
            <Dimmer
              active={true}
            >
              <DimmerInner
                active={true}
              >
                <Ref
                  innerRef={
                    Object {
                      "current": <div
                        class="ui active transition visible dimmer"
                        style="display: flex;"
                      >
                        <div
                          class="content"
                        >
                          <div
                            class="ui large text loader"
                          >
                            Loading
                          </div>
                        </div>
                      </div>,
                    }
                  }
                >
                  <RefFindNode
                    innerRef={
                      Object {
                        "current": <div
                          class="ui active transition visible dimmer"
                          style="display: flex;"
                        >
                          <div
                            class="content"
                          >
                            <div
                              class="ui large text loader"
                            >
                              Loading
                            </div>
                          </div>
                        </div>,
                      }
                    }
                  >
                    <div
                      className="ui active transition visible dimmer"
                      onClick={[Function]}
                    >
                      <div
                        className="content"
                      >
                        <Loader
                          size="large"
                        >
                          <div
                            className="ui large text loader"
                          >
                            Loading
                          </div>
                        </Loader>
                      </div>
                    </div>
                  </RefFindNode>
                </Ref>
              </DimmerInner>
            </Dimmer>
            <div
              className="form-group"
              style={
                Object {
                  "display": "block",
                }
              }
            >
              <div
                className="input-div d-flex justify-content-center"
                style={
                  Object {
                    "marginBottom": "1rem",
                  }
                }
              >
                <Input
                  className="form-control-input"
                  name="email"
                  onChange={[Function]}
                  placeholder="Email"
                  required="required"
                  type="email"
                  value=""
                >
                  <input
                    className="form-control-input"
                    name="email"
                    onChange={[Function]}
                    placeholder="Email"
                    required="required"
                    type="email"
                    value=""
                  />
                </Input>
              </div>
              <div
                className="input-div d-flex justify-content-center"
              >
                <Input
                  className="form-control-input"
                  name="password"
                  onChange={[Function]}
                  pattern="(?=.*[a-z]).{6,}"
                  placeholder="Password"
                  required="required"
                  title="Must be 6 characters or more and contain at least lowercase letter"
                  type="password"
                  value=""
                >
                  <input
                    className="form-control-input"
                    name="password"
                    onChange={[Function]}
                    pattern="(?=.*[a-z]).{6,}"
                    placeholder="Password"
                    required="required"
                    title="Must be 6 characters or more and contain at least lowercase letter"
                    type="password"
                    value=""
                  />
                </Input>
              </div>
            </div>
            <div
              style={
                Object {
                  "marginTop": ".9235rem",
                }
              }
            >
              <Button
                className="btn-dark"
                type="submit"
                value="WELCOME BACK"
              >
                <button
                  className="btn btn-dark"
                  type="submit"
                >
                  WELCOME BACK
                </button>
              </Button>
            </div>
            <p
              className="f-20 text-grey text-center"
              style={
                Object {
                  "marginTop": ".9235rem",
                }
              }
            >
              Don't have an account? Register
               
              <Link
                className="text-blue"
                to="/register"
              >
                <a
                  className="text-blue"
                  href="/register"
                  onClick={[Function]}
                >
                  here
                </a>
              </Link>
            </p>
          </form>
        </div>
      </div>
    </Login>
  </Router>
</BrowserRouter>
`;
